### ISSUE TITLE
Support standard, category, sniff, and error codes in `--sniff` and `--exclude`

### ISSUE THREAD

**Issue by yCodeTech** [PHPCSStandards/PHP_CodeSniffer#1497]
## Is your feature request related to a problem?

I am the current maintainer of the [PHP Sniffer & Beautifier vscode extension]( https://github.com/valeryan/vscode-phpsab), which displays the sniff errors in VSCode's Problem's panel as well as an in-editor on-hover popup. 

I'd like to implement some functionality that uses VSCode's CodeActions and QuickFix to fix that particular error code that VSCode reports. This would use the `--sniffs` argument to limit the scan to that error code, which isn't currently possible.

So allowing standard, category, sniff, and error codes to be used in `--sniffs` and `--exclude` arguments in both `phpcs` and `phpcbf` would be a favourable addition.

## Describe the solution you'd like

Support the usage of any form of the 4-part codes for the `--sniffs` and `--exclude` arguments. Example below shows usage of the arguments with all 4 parts in `phpcbf`, and would also apply to `phpcs` too.

```bash
phpcbf file.php --sniffs=Squiz
phpcbf file.php --sniffs=Squiz.Functions
phpcbf file.php --sniffs=Squiz.Functions.FunctionDeclarationArgumentSpacing # current implementation
phpcbf file.php --sniffs=Squiz.Functions.FunctionDeclarationArgumentSpacing.SpacingAfterOpen

phpcbf file.php --exclude=Squiz
phpcbf file.php --exclude=Squiz.Functions
phpcbf file.php --exclude=Squiz.Functions.FunctionDeclarationArgumentSpacing # current implementation
phpcbf file.php --exclude=Squiz.Functions.FunctionDeclarationArgumentSpacing.SpacingAfterOpen
```

## Additional context (optional)

[gsherwood's comment](https://github.com/squizlabs/PHP_CodeSniffer/pull/2646#issuecomment-541255248) original intent was to eventually allow all 4 parts in the CLI:

> I intend on allowing Standard, Category, Sniff, and message codes on the CLI at some point in the future so that the feature set lines up with how rulesets work...

I think this is a good idea to implement, because it will bring functional parity to the CLI in-line with the rulesets, and allow users more control over the CLI arguments.

- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/HEAD/.github/CONTRIBUTING.md) and this is not a [support question](https://github.com/PHPCSStandards/PHP_CodeSniffer/discussions).
- [ ] I intend to create a pull request to implement this feature.

**Comment by jrfnl** [comment]
@yCodeTech I like the idea. Have you thought through the pitfalls though ? For example, it will be impossible to validate the error codes provided on the command line, which will mean that PHPCS will need to silently ignore any provided error code which is invalid,

In my opinion, these kind of limitations will need to be thought through properly and the consequences discussed before a PR for a feature like this can be opened. They should also help clarify if the feature is worth the effort given the limitations it will have.

_As sidenotes:_
* _Please be aware that the parts of the codebase which would need to be adjusted to make this work are barely covered by tests, so the chances of breaking things are high. It would be preferable to ensure the current functionality is covered by tests first, before making adjustments to that code._
* _Also please be aware that implementing new features like this is not a priority, while PHPCS is behind in supporting PHP syntaxes._

**Comment by yCodeTech** [comment]
@jrfnl To be fair, I didn't think about the pitfalls.

>  For example, it will be impossible to validate the error codes provided on the command line, which will mean that PHPCS will need to silently ignore any provided error code which is invalid,

I do believe this is also true for specifying error codes in the rulesets? It just silently ignores them, while the 3-part codes are validated.

After having a look at `File.php`, correct me if I'm wrong, but I don't think the error code is being stored/cached anywhere? So my first thought would be to collect all the error codes associated with the sniff in some sort of (global?) mapping. The `addMessage` method is probably the best place to grab the passed error codes and add it to the collection. This will then allow whatever class/method to access the collection of stored error codes per sniff, and validate them both on the CLI and the rulesets.

That's just how I imagine it to be integrated, I couldn't imagine it being much of a breaking change(?). I'm also not sure about any limitations/pitfalls/consequences of this. I guess maybe if all the standards aren't "read" until a certain standard is being used or specifically requested in the CLI/ruleset, then my idea can't cache the error codes. But without knowledge of how it actually all works together, to me it seems relatively straightforward and will help the integrity of sniffs to error codes. But I'm not an expert on the code as you are, and maybe you can fill me in on some lack of understanding where necessary?

**Comment by jrfnl** [comment]
> After having a look at `File.php`, correct me if I'm wrong, but I don't think the error code is being stored/cached anywhere? So my first thought would be to collect all the error codes associated with the sniff in some sort of (global?) mapping. The `addMessage` method is probably the best place to grab the passed error codes and add it to the collection. This will then allow whatever class/method to access the collection of stored error codes per sniff, and validate them both on the CLI and the rulesets.

@yCodeTech I appreciate your thoughts about this, but don't understand how you envision this would work in practice ?

The `addMessage()` method is only called when a sniff throws a warning/error, while the CLI flags and ruleset need to be interpreted well before that time to even know what sniffs to load, so anything involving the `addMessage()` method would not solve the validation problem.

A couple of things come to mind straight away:
* A lot of sniffs use dynamic error codes, so even if we'd tokenize the sniff files ahead of time and look for the function calls to `addError()` and friends, we wouldn't be able to figure out all the error codes in use by such a sniff.
    This would also incur a hefty performance penalty.
* If done at runtime, it's too late to flag invalid codes and any such "collecting" will rarely, if ever, be complete unless all error codes possible would be hit in a run, which generally only happens when testing sniffs.
* If we'd demand that sniffs provide all possible error codes via a method which the framework could call while processing the CLI and rulesets, that would be a huge breaking change and would add a significant maintenance burden for every single PHPCS standard.

**Comment by yCodeTech** [comment]
@jrfnl 
> The addMessage() method is only called when a sniff throws a warning/error, while the CLI flags and ruleset need to be interpreted well before that time to even know what sniffs to load, so anything involving the addMessage() method would not solve the validation problem.

I was only looking at where the error codes came from (using `addError`, etc.), that's all. With the codes not defined anywhere else, the only thing I could think of was working backwards from the error code source, which in hindsight is not how it works.

> A lot of sniffs use dynamic error codes

Oh, I didn't know that. I know the error messages can be dynamic, but I thought the error codes were static strings.

> if we'd tokenize the sniff files ahead of time and look for the function calls to addError()

I guess that's what I was thinking of when I wrote my idea. Tokenizing the sniff files to figure out the error codes. But you're right, if the error codes are dynamic, then validation wouldn't work. It would only work if all error codes were static strings. But then it'd be then performance issues as you're stated.

> If done at runtime, it's too late to flag invalid codes

Fair point.

> If we'd demand that sniffs provide all possible error codes via a method which the framework could call while processing the CLI and rulesets, that would be a huge breaking change and would add a significant maintenance burden for every single PHPCS standard.

I guess this would be the ideal solution - a massive refactor of how it works to ensure all error codes could be accounted for. But it would indeed have massive implications on standard developers, and wouldn't be feasible for such a big refactor. 

Though having said that, it would also help find and document all sniffs and their error codes either via 3rd party software or internally via the wiki or future website. Because at its core, this is the same issue as what other users have already pinpointed in many issues - how to find out what error codes a sniff contributes. Especially if the standard developer didn't bother to document them on their repository readme. I've had to go through entire 3rd party and phpcs internally-supported standards to just find error codes - it does seem to be a consistent problem. So making sniffs declare all possible error codes, which would then help with my collections idea, would certainly help to search and document the error codes, whilst also contributing to the validation.

Going back to my initial feature request, the fact that invalid 4-part codes already fail silently when used in the rulesets, so I don't really see much difference in allowing them to fail silently in the CLI either. I understand your point and I agree that validating the error codes would be better, but allowing them in the rulesets and not in the CLI seems strange to me. It would make more sense (to me) to bring the CLI inline with the rulesets, and allow error codes to be included/excluded, for the time being while a suitable solution for validation is found.

I understand that you're busy and not looking for big changes or feature implementations right now. But for the future, I wouldn't mind helping to implement a solution when it's suitable and viable. Whatever the solution, I do think it will be worth it in the long run, for both validating the error codes in the CLI and rulesets, and also allowing the codes to be more visible and easily documented.
